lib/Lower/ConvertExpr.cpp

```diff
@@ -795,66 +798,89 @@ class ExprLowering {
-               "optional argument requires explicit interface");
-        const auto *expr = arg->UnwrapExpr();
-        assert(expr && "assumed type argument requires explicit interface");
-        if (const auto *sym =
-                Fortran::evaluate::UnwrapWholeSymbolDataRef(*expr)) {
-          mlir::Value argRef = symMap.lookupSymbol(*sym);
-          assert(argRef && "could not get symbol reference");
-          argTypes.push_back(argRef.getType());
-          operands.push_back(argRef);
-        } else {
-          // TODO create temps for expressions
-          TODO();
-        }
+    }
+    // Let the intrinsic library lower the intrinsic procedure call
+    llvm::StringRef name{intrinsic.name};
+    return intrinsics.genval(getLoc(), builder, name, resultType[0], operands);
+  }
+
+  mlir::Value genProcedureRef(const Fortran::evaluate::ProcedureRef procRef,
+                              mlir::ArrayRef<mlir::Type> resultType) {
+    if (const auto *intrinsic{procRef.proc().GetSpecificIntrinsic()}) {
+      return genIntrinsicRef(procRef, *intrinsic, resultType[0]);
+    }
+    // Implicit interface implementation only
+    // TODO: Explicit interface, we need to use Characterize here,
+    // evaluate::IntrinsicProcTable is required to use it.
+    llvm::SmallVector<mlir::Type, 2> argTypes;
+    llvm::SmallVector<mlir::Value, 2> operands;
+    // Logical arguments of user functions must be lowered to `fir.logical`
+    // and not `i1`.
+    auto restorer = Fortran::common::ScopedSet(genLogicalAsI1, false);
+    for (const auto &arg : procRef.arguments()) {
+      if (!arg.has_value())
+        TODO(); // optional arguments
+      const auto *expr = arg->UnwrapExpr();
+      if (!expr)
+        TODO(); // assumed type arguments
+      if (const auto *sym =
+              Fortran::evaluate::UnwrapWholeSymbolDataRef(*expr)) {
+        mlir::Value argRef = symMap.lookupSymbol(*sym);
+        assert(argRef && "could not get symbol reference");
+        argTypes.push_back(argRef.getType());
+        operands.push_back(argRef);
+      } else {
+        // TODO create temps for expressions
+        TODO();
       }
-      mlir::Type resultType = converter.genType(TC, KIND);
-      mlir::FunctionType funTy =
-          mlir::FunctionType::get(argTypes, resultType, builder.getContext());
-      mlir::FuncOp func = getFunction(applyNameMangling(funRef.proc()), funTy);
-      auto call = builder.create<mlir::CallOp>(getLoc(), func, operands);
-      // For now, Fortran return value are implemented with a single MLIR
-      // function return value.
-      assert(call.getNumResults() == 1 &&
-             "Expected exactly one result in FUNCTION call");
-      return call.getResult(0);
     }
+    mlir::FunctionType funTy =
+        mlir::FunctionType::get(argTypes, resultType, builder.getContext());
+    auto funName = applyNameMangling(procRef.proc());
+    ;
+    [[maybe_unused]] mlir::FuncOp func = getFunction(funName, funTy);
```

Comment:
Is this ever going to be used? I'm guessing we want some side-effect of calling `getFunction()`?

Reply:
Right, I removed the var and just kept the call.